### PR TITLE
Support Smallrye Config property expression expansion for @RolesAllowed value

### LINKED ISSUE #25245: EL Placeholder support in @RolesAllowed annotation
### Describe the bug

If we use an EL placeholder in @RolesAllowed annotation value, to make a reference to a config property, the placeholder is not resolved.
Example:

**MyClass.java**

```
@RolesAllowed("myrole_${role.suffix}")
public class MyClass
```

**application.yml**
```
role:
  suffix: someValue
```

This is something that works with Jakarta EE on EAP/Wildfly so I expected it to work also with Quarkus.

### Expected behavior

@RolesAllowed annotation value resolved to "myrole_someValue"

### Actual behavior

@RolesAllowed annotation value remains "myrole_${role.suffix}"

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by michalvavrik**
closes: #25245

Support SmallRye Config property expressions expansion for the `@RolesAllowed` annotation value.

### CODE REVIEW

File: docs/src/main/asciidoc/security-authorization-of-web-endpoints-reference.adoc [hunk 1]
```diff
@@ -263,13 +263,23 @@ public class SubjectExposingResource {
...
+        return name;
+    }
 }
 ----
 <1> This `/subject/secured` endpoint requires an authenticated user that has been granted the role "Tester" through the use of the `@RolesAllowed("Tester")` annotation.
 <2> The endpoint obtains the user principal from the JAX-RS `SecurityContext`. This will be non-null for a secured endpoint.
 <3> The `/subject/unsecured` endpoint allows for unauthenticated access by specifying the `@PermitAll` annotation.
 <4> This call to obtain the user principal will return null if the caller is unauthenticated, non-null if the caller is authenticated.
 <5> The `/subject/denied` endpoint disallows any access regardless of whether the call is authenticated by specifying the `@DenyAll` annotation.
+<6> The `@RolesAllowed` annotation value supports <<config-reference#property-expressions,Property Expressions>> including default values and nested Property Expressions.
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-authorization-of-web-endpoints-reference.adoc:282]:
```suggestion
<6> The `@RolesAllowed` annotation value is set to the value of the `admin` configuration property, see <<config-reference#property-expressions,Property Expressions>> for more information.
```

Comment (sberyozkin) [reply docs/src/main/asciidoc/security-authorization-of-web-endpoints-reference.adoc:282]: I believe in this example, it has to be a concrete piece of info, can make sense IMHO to have a separate dedicated section with a  more technical overview

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java [hunk 2]
```diff
@@ -176,9 +176,7 @@ public Uni<? extends SecurityIdentity> apply(SecurityIdentity securityIdentity)
                                 public void accept(SecurityIdentity identity, Throwable throwable, Boolean aBoolean) {
                                     if (identity != null) {
                                         //when the result is evaluated we set the user, even if it is evaluated lazily
-                                        if (identity != null) {
```
[the remark this answers is not in the document]
Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java:179]: It is not, but it's so super tiny it's not worth separate PR and it's clearly surplus to requirements. I can feel you don't like it, so let's me remove this from the PR...

File: integration-tests/elytron-resteasy-reactive/src/main/resources/application.properties [hunk 3]
```diff
@@ -7,3 +7,4 @@ quarkus.security.users.embedded.users.poul=poul
 quarkus.security.users.embedded.roles.poul=interns
 quarkus.security.users.embedded.plain-text=true
 quarkus.http.auth.basic=true
+employees-config-property=employees
```
Comment (sberyozkin) [inline integration-tests/elytron-resteasy-reactive/src/main/resources/application.properties:10]: I'm not sure I see what is happening here, the authenticated user must have an `interns` role, how does `employees-config-property=employees` align with it ?

File: docs/src/main/asciidoc/security-authorization-of-web-endpoints-reference.adoc (hunk 1 above)
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-authorization-of-web-endpoints-reference.adoc:282]: I agree with dedicated section, just wan't sure it's important enough. I'll write something.

File: integration-tests/elytron-resteasy-reactive/src/main/resources/application.properties (hunk 3 above)
Comment (michalvavrik) [reply integration-tests/elytron-resteasy-reactive/src/main/resources/application.properties:10]: please check this line @sberyozkin https://github.com/quarkusio/quarkus/blob/main/integration-tests/elytron-resteasy-reactive/src/main/resources/application.properties#L3

Comment (michalvavrik) [reply integration-tests/elytron-resteasy-reactive/src/main/resources/application.properties:10]: Authenticated user that needs access to `employee` needs `employees` role.

Comment (sberyozkin) [reply integration-tests/elytron-resteasy-reactive/src/main/resources/application.properties:10]: Oh, sorry, I forgot that the actual enhancement is about being able to vary the roles allowed value (for a sec I thought that property was kind of disconnected from the actual namespaced authentication configuration, ex, for the basic auth - but it is really about controlling dynamically users in which role can access this method)

File: docs/src/main/asciidoc/security-authorization-of-web-endpoints-reference.adoc (hunk 1 above)
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-authorization-of-web-endpoints-reference.adoc]: I completely re-wrote that, let see please and thank you.

File: docs/src/main/asciidoc/security-authorization-of-web-endpoints-reference.adoc [hunk 4]
```diff
@@ -273,6 +273,75 @@ public class SubjectExposingResource {
 
 CAUTION: Please refer to the xref:security-built-in-authentication-support-concept.adoc#proactive-authentication[Proactive Authentication] section of the Built-In Authentication Support guide if you plan to use standard security annotations on IO thread.
 
+The `@RolesAllowed` annotation value supports <<config-reference#property-expressions,Property Expressions>> including default values and nested Property Expressions.
+Configuration properties used with the annotation are resolved at runtime and supports configuration profiles.
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-authorization-of-web-endpoints-reference.adoc:277]:
```suggestion
Configuration properties used with the annotation are resolved at runtime.
```

Comment (sberyozkin) [reply docs/src/main/asciidoc/security-authorization-of-web-endpoints-reference.adoc:277]: I thought the note about configuration profile does not clarify the message that these properties are resolved at runtime - the profiles are part of the Config system

**Review by sberyozkin — APPROVED** [review]
@michalvavrik This is a nice feature to have, thanks, please look at the last single comment.
Let @stuartwdouglas to have a quick look if he gets a chance, I'll merge tomorrow, Stuart, it can be tweaked as always if you notice something

File: docs/src/main/asciidoc/security-authorization-of-web-endpoints-reference.adoc (hunk 4 above)
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-authorization-of-web-endpoints-reference.adoc]: Looking at it now, I think examples below with 2 configuration profiles are enough. I removed the note, thank you.

**Review by radcortez — COMMENTED** [review]
Just a few thoughts:

Did you consider generating a configuration source with generated keys and values with the property expressions? The source could be registered with the Config system, and you would get all features available from Config and not only expansion.

Another idea could be to support configuration properties, and instead of the expansion, we do a complete config lookup (which, of course, also supports expansion).

**Comment by michalvavrik** [comment]
> Just a few thoughts:
> 
> Did you consider generating a configuration source with generated keys and values with the property expressions? The source could be registered with the Config system, and you would get all features available from Config and not only expansion.
> 
> Another idea could be to support configuration properties, and instead of the expansion, we do a complete config lookup (which, of course, also supports expansion).

I didn't think of that, I actually took identical approach as Scheduler extension, but I like that very much. @sberyozkin I think Roberto's suggestion is going to be less prone to (future) compatibility issues and and we get Smallrye Config improvements for free with next Smallrye Config versions. I'll have a look at this, please hold with merging till I investigate how to do it, it shouldn't take long. Thank you

**Comment by michalvavrik** [comment]
Okay, I found a way, I'll do it tomorrow.

**Review by stuartwdouglas — APPROVED** [review]

**Comment by sberyozkin** [comment]
Thanks @radcortez, @michalvavrik Sure, take your time please.

**Comment by radcortez** [comment]
Great!

Here are some specific ideas. Feel free to take them or discard them :)

```java
@RolesAllowed("security.role")
void single() {
}

@RolesAllowed("security.roles")
void multiple() {
}

@RolesAllowed("security.expressiom")
void expression() {
}

@RolesAllowed("actualRole") // We cannot find in in Config, so we use the value as is
void noConfig() {
}
```

```properties
security.role=role
security.roles=role1,role2
#or
security.role[0]=role1
security.role[1]=role2
security.expression=role_${number}
number=1
```

**Comment by michalvavrik** [comment]
> generating a configuration source with generated keys and values with the property expressions

I used your suggestion to generate a configuration source with generated keys and values with the property expressions and also to use `RunTimeConfigBuilderBuildItem`. There was something that surprised me - `RunTimeConfigBuilderBuildItem` don't work properly with classes generated through `GeneratedClassGizmoAdaptor` (it leads to class not found ex), while `StaticInitConfigSourceProviderBuildItem` does. For that reason I decided to collect properties for config source at STATIC INIT, no blocker.

And I kept previous behavior - I don't consider all `@RolesAllowed` to be config properties. For one, I think that would be breaking change and also using config expressions is something that users know from EAP.

Thank you very much for help

**Comment by radcortez** [comment]
> There was something that surprised me - `RunTimeConfigBuilderBuildItem` don't work properly with classes generated through `GeneratedClassGizmoAdaptor` (it leads to class not found ex), while `StaticInitConfigSourceProviderBuildItem` does.

Yes, I'm aware of this. The issue is that the builders are expected to be available in the bootstrap classloader (the `isApplicationClass` flag in GeneratedClassGizmoAdaptor)`. I've already changed this in https://github.com/quarkusio/quarkus/pull/29842 to use the context ClassLoader.

**Comment by sberyozkin** [comment]
Thanks Michal, Roberto

**Comment by geoand** [comment]
@sberyozkin @michalvavrik this seems to have broken `main`.

For example `io.quarkus.it.resteasy.reactive.elytron.FruitResourceIT` is currently failing with:

```
ERROR: Failed to start application (with profile [prod])
java.lang.ClassNotFoundException: io.quarkus.security.runtime.QuarkusSecurityRolesAllowedConfigBuilder
        at java.base@17.0.5/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:52)
        at java.base@17.0.5/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base@17.0.5/java.lang.ClassLoader.loadClass(ClassLoader.java:132)
        at io.quarkus.runtime.configuration.ConfigUtils.configBuilder(ConfigUtils.java:215)
        at io.quarkus.runtime.generated.Config.readConfig(Unknown Source)
        at io.quarkus.deployment.steps.RuntimeConfigSetup.deploy(Unknown Source)
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:108)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:124)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
```

**Comment by geoand** [comment]
https://github.com/quarkusio/quarkus/pull/30220 should fix the issue

**Pull request merged**